adc/consumer.py

```diff
@@ -66,7 +66,8 @@ def mark_done(self, msg: confluent_kafka.Message):
         continuously synchronize this information with Kafka so that the stream can be
         resumed from this point in the future.
         """
-        self._consumer.store_offsets(msg)
+        self._consumer.commit(msg, asynchronous=True)
+        self._consumer.poll(1e-3) # wait at most a millisecond
```

Comment:
I'm a bit concerned about the performance implications of this. I'm also worried that the `poll()` is not doing what is expected, since my understanding is that `poll()` can potentially return a message so that this may be the cause of lost messages in rare cases if a message gets returned before the timeout is hit.

Sorry, I had forgotten what the motivation for this PR was, but what is this solving that wasn't present before?

Reply:
I think performance can only become an issue for message rates above 100 messages per second. However, your point about `poll()` possibly returning a message is a serious problem, I think. I will investigate again; unfortunately, my recollection of why I added that was that I needed it to force the offsets to actually be committed, and if that is the case I will need to find some other way to do that without risking getting an extra message back. I agree that the current form may just lose a message, so we clearly can't merge this as-is.

Reply:
Okay, I see. It looks like there's at least two potential scenarios where messages may not be committed even if they're stored via `store_offsets()`:

1. A `KafkaException` gets raised somewhere in the middle of processing messages within `stream()`
2. Some other error condition is raised in a downstream client.

In both cases, I believe it will only happen if `close()` isn't called or if `close()` is called but `enable.auto.commit` is set to False. Is that your understanding as well?

Reply:
I'd like to propose an alternative and want to hear your thoughts about it. Instead of modifying `mark_done()`, we can add an explicit `commit()` in two places:

* In `close()` just before the consumer is closed.
* Wrapped in a try/finally loop when messages are yielded in the private stream methods, such that finally calls `commit()` and regardless of whether a message needs to be raised with an error, it'll hit the finally part and do an explicit commit.

Within the context manager in the hop client, `close()` should always be called regardless so the offsets should be committed in this case in the cases where it was not before.

How that sound reasonable or are there problems with this implementation?

Reply:
Using `poll` was just dangerous, so I've started by taking it out. I think the key thing we should look at instead is whether we want to call `commit` with `asynchronous` set or not. With it set to false, as it is currently, it may not fully take effect (or return an error on failure) until `poll` is called. I think we may want to use `asynchronous=False` in at least some cases to force `commit` to succeed or fail up front. it seems like since autocommit corresponds to being a bit lax about whether the user code has truly fully processed the message or no anyway, we might want it to use `asynchronous=True` to let things be a little more efficient if possibly less safe and precise. In that context it might also make sense to put a final `commit(asynchronous=True)` into `close` (However, there's the slightly tricky point that I'm not sure what the other aurgments should be; at that point we don't have any specific message to commit, and the [smantics of specifying neither a message nor a partition list to commit](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#ab96539928328f14c3c9177ea0c896c87) aren't currently clear to me). Meanwhile, for when the user is not using `autocommit`, I think we would want to let them call `mark_done` explicitly, and have it use `commit(asynchronous=False)` to ensure that exactly what the user asked for is being done. 

I'm not sure I follow why we would want the `commit` in a `try. . . finally` construct in the `stream*` methods. Is the idea to move the `commit` to happening after the message has been yielded, when autocommit is on? That seems like it would be a slight improvement, in that if the use is requesting another message, they have probably finished with the last one. However, we still won't know for sure, so i'm not sure if there's much point, since autocommit will always be guessing a bit about the user's intentions, and so can't ever really be totally precise. I also don't understand why we would want to commit if in fact an error occured. In that case, don't we know that the message cannot have be handled, so we defintiely do not want to commit?

Reply:
> Using `poll` was just dangerous, so I've started by taking it out. I think the key thing we should look at instead is whether we want to call `commit` with `asynchronous` set or not. With it set to false, as it is currently, it may not fully take effect (or return an error on failure) until `poll` is called. I think we may want to use `asynchronous=False` in at least some cases to force `commit` to succeed or fail up front. it seems like since autocommit corresponds to being a bit lax about whether the user code has truly fully processed the message or no anyway, we might want it to use `asynchronous=True` to let things be a little more efficient if possibly less safe and precise. In that context it might also make sense to put a final `commit(asynchronous=True)` into `close` (However, there's the slightly tricky point that I'm not sure what the other aurgments should be; at that point we don't have any specific message to commit, and the [smantics of specifying neither a message nor a partition list to commit](https://docs.confluent.io/platform/current/clients/librdkafka/html/rdkafka_8h.html#ab96539928328f14c3c9177ea0c896c87) aren't currently clear to me). Meanwhile, for when the user is not using `autocommit`, I think we would want to let them call `mark_done` explicitly, and have it use `commit(asynchronous=False)` to ensure that exactly what the user asked for is being done.

My understanding with calling `commit()` without any arguments is that is uses the last stored offsets to `store_offsets()` (or whatever the last message has come out of `poll()` or `consume()` if `enable.auto.offset.store` is enabled. When you'd use `asynchronous=False` seems reasonable to me when you call `commit()`. When `close` is being called, do you mean calling `commit` synchronously instead to force it to succeed or fail up front or did I miss something? The `mark_done` suggestion sounds reasonable in terms of what the user expects, I'm just worried about the performance implication of that if it's a problem, and there's a possibility to achieve effectively the same behavior without it. Maybe those concerns are unfounded, however.

> I'm not sure I follow why we would want the `commit` in a `try. . . finally` construct in the `stream*` methods. Is the idea to move the `commit` to happening after the message has been yielded, when autocommit is on? That seems like it would be a slight improvement, in that if the use is requesting another message, they have probably finished with the last one. However, we still won't know for sure, so i'm not sure if there's much point, since autocommit will always be guessing a bit about the user's intentions, and so can't ever really be totally precise. I also don't understand why we would want to commit if in fact an error occured. In that case, don't we know that the message cannot have be handled, so we defintiely do not want to commit?

So maybe this isn't the best route to take, but my idea was to use `store_offsets()` within `mark_done()` and then when the batch of messages are consumed, finally commit those offsets when the batch of messages have been processed or an error was encountered. In the case where the an error occurred, mark_done() was never called to commit the offsets, so an empty call (except with the `asynchronous` kwarg) would not commit the message with the error contained, but everything up to that last message. My thought process was to avoid the performance hit with calling `commit()` repeatedly but still ensure that `commit()` was always going to be called at some point (whether that's actually the case with what I mentioned above) before exiting so that the behavior is still consistent with calling `commit()` on every message.

Reply:
> When close is being called, do you mean calling commit synchronously instead to force it to succeed or fail up front or did I miss something?

That was my thinking, yes. 

> I'm just worried about the performance implication of that if it's a problem, and there's a possibility to achieve effectively the same behavior without it. Maybe those concerns are unfounded, however.

My figuring is that in the worst case this would about halve throughput, since it would introduce an extra roundtrip communication with the broker per message processed. Some testing might be in order, but my hope is that we should still be able to sustain a pretty good rate even in this case. Using `asynchronous=True` will presumably reduce the overhead by batching the communications, at the cost of reducing precision and slightly widening the timing in which we might fail to keep the broker informed of our latest state. THis feels to me like a typical distributed systems tradeoff. 

> my idea was to use store_offsets() within mark_done() and then when the batch of messages are consumed, finally commit those offsets when the batch of messages have been processed or an error was encountered.

I see. I think that's taking the same tradeoff a bit further. The danger, as I see it, is that during the dealy while accumulating a batch the client might just die entirely due to an unhandleable error (process killed, whole machine crashes, etc.). That is of course not very likely and we can never be completely immune to it, just reduce the time windows in which it would cause a loss of information and reduce how much information can be lost. 

> My thought process was to avoid the performance hit with calling commit() repeatedly but still ensure that commit() was always going to be called at some point

That makes sense, and is basically the same thing I was after as well, I was just figuring that `commit(asynchronous=True)` would also significantly mitigate the overhead (though I agree it must have higher overhead than several calls to `store_offsets` per `commit`). I guess I'll try to do some benchmarking to figure out to what degree any of this makes a difference.

Reply:
Those are all really good points. I don't have a good sense of what tradeoffs make sense here, so I'll most likely rally behind whatever you end up comfortable with. Just let me know when you're ready and I'll take another look.

Reply:
Some simple measurements, reading 1101 messages from a topic:

Fully synchronous commit for each message:
```
 285551 function calls (280188 primitive calls) in 92.132 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1101   88.095    0.080   88.095    0.080 {method 'commit' of 'cimpl.Consumer' objects}
       13    2.944    0.226    2.944    0.226 {method 'consume' of 'cimpl.Consumer' objects}
        1    0.752    0.752    0.753    0.753 {method 'list_topics' of 'cimpl.Consumer' objects}
     1101    0.033    0.000    0.033    0.000 {built-in method builtins.print}
        1    0.016    0.016    0.016    0.016 consumer.py:19(__init__)
```
(12 messages per second)

Asynchronous commit for each message:

```
 285545 function calls (280182 primitive calls) in 18.182 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   14.499   14.499   14.499   14.499 {method 'close' of 'cimpl.Consumer' objects}
       13    2.786    0.214    2.786    0.214 {method 'consume' of 'cimpl.Consumer' objects}
        1    0.654    0.654    0.654    0.654 {method 'list_topics' of 'cimpl.Consumer' objects}
        1    0.015    0.015    0.015    0.015 consumer.py:19(__init__)
```
(62 messages per second)

store_offsets per message, asynchronous commit per batch, batch size 100, autocommit interval 100 ms:
```
 285566 function calls (280203 primitive calls) in 3.619 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       13    2.578    0.198    2.578    0.198 {method 'consume' of 'cimpl.Consumer' objects}
        1    0.639    0.639    0.639    0.639 {method 'list_topics' of 'cimpl.Consumer' objects}
        1    0.158    0.158    0.158    0.158 {method 'close' of 'cimpl.Consumer' objects}
        1    0.015    0.015    0.015    0.015 consumer.py:19(__init__)
```
(369 messages per second)

store_offsets per message, asynchronous commit per batch, batch size 10, autocommit interval 100 ms:
```
 286033 function calls (280670 primitive calls) in 4.320 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      113    3.328    0.029    3.328    0.029 {method 'consume' of 'cimpl.Consumer' objects}
        1    0.677    0.677    0.677    0.677 {method 'list_topics' of 'cimpl.Consumer' objects}
        1    0.087    0.087    0.087    0.087 {method 'close' of 'cimpl.Consumer' objects}
        1    0.016    0.016    0.016    0.016 consumer.py:19(__init__)
```
(302 messages per second)

store_offsets per message, asynchronous commit per batch, batch size 1, autocommit interval 100 ms:
```
 290008 function calls (284644 primitive calls) in 21.134 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1   17.703   17.703   17.703   17.703 {method 'close' of 'cimpl.Consumer' objects}
     1118    2.518    0.002    2.518    0.002 {method 'consume' of 'cimpl.Consumer' objects}
        1    0.670    0.670    0.670    0.670 {method 'list_topics' of 'cimpl.Consumer' objects}
        1    0.015    0.015    0.015    0.015 consumer.py:19(__init__)
```
(53 messages per second)


Observations from this:
- Being very safe and doing synchronous commits per message is very slow. This is disappointing, but it's just what Kafka/librdkafka gives us, and I think we probably still need to offer this as an option for anyone who really wants it. 
- Doing an asynchronous commit per message works disappointingly poorly. It's faster than synchronous commits, but not much, and it appears that redundant queued commit requests are not coalesced, which can lead to the situation such that that calling `commit` more often is actually worse, as a large number of commits get saved up to execute during `close`, and then take a long time to do so (apparently serially). This can lead to a larger time window in which a disruption will cause commits to be lost, rather than a smaller one. 
- Using `store_offsets` generally has reasonable, if not amazing, throughput, but periodic explicit calls are, as you have noted, still required. Tying that to the batch size seems like the most practical thing to do. Unfortunately, with a small enough batch size this does devolve into the pathological behavior of too many asynchronous commits.

Reply:
Interesting results and nice to see some quantitative results attached to these various cases. Based on this, what do you suggest moving forward?

Reply:
The scheme currently implemented in this branch, with autocommit using `store_offsets` per mesage and asynchronous `commit` per batch, and users not using autocommit having the choice between `store_offsets` and synchronous `commit` is the best I've been able to come up with. Between having the option to tune the batch size in the first case, and to choose when to commit in the latter, I think this gives users the full range of flexibility.  Documenting how and when to use these options still needs to be done, though I was thinking that would be part of a future PR to write how-to guides for advanced use of our API.

Reply:
Agreed, that sounds reasonable to me as well.